internal/authz/github/authz.go

```diff
@@ -29,7 +31,36 @@ func NewAuthzProviders(
 		}
 	}
 
+	// Auth providers (i.e. login mechanisms)
+	githubAuthProviders := make(map[string]struct{})
+	for _, p := range authProviders {
+		if p.Github != nil {
+			var id string
+			ghURL, err := url.Parse(p.Github.Url)
+			if err != nil {
+				// error reporting for this should happen elsewhere, for now just use what is given
+				id = p.Github.Url
+			} else {
+				// use codehost normalized URL as ID
+				ch := extsvc.NewCodeHost(ghURL, p.Github.Type)
+				id = ch.ServiceID
+			}
+			githubAuthProviders[id] = struct{}{}
+		}
+	}
+
 	for _, p := range ps {
+		// Permissions require a corresponding GitHub oauth provider. Without one, repos
+		// with restricted permissions will not be visible to non-admins.
+		if _, exists := githubAuthProviders[p.ServiceID()]; !exists {
+			warnings = append(warnings,
+				fmt.Sprintf("Did not find authentication provider matching %q. "+
+					"Check the [**site configuration**](/site-admin/configuration) to "+
+					"verify an entry in [`auth.providers`](https://docs.sourcegraph.com/admin/auth) exists for %s.",
+					p.ServiceID(), p.ServiceID()))
```

Comment:
```suggestion
				fmt.Sprintf("Did not find authentication provider matching %[1]q. "+
					"Check the [**site configuration**](/site-admin/configuration) to "+
					"verify an entry in [`auth.providers`](https://docs.sourcegraph.com/admin/auth) exists for %[1]s.",
					p.ServiceID()))
```

Also why the second parameter is `%s` not `%q`? 🤔

Reply:
This is copied over from the GitLab errors message. The first one is quote, I think, because we're referring to a 'value', whereas the second one is identifying this permissions provider (which uses no quote, i.e. `"GitHub config for %s was invalid: %s"`

Reply:
Going to merge for now - happy to address in a followup!

Additional thought - I think the error reporting has some room for improvement here, it can be confusing to figure out which code host the problem is in. Maybe we can get some more user-friendly metadata e.g. `DisplayName`

Reply:
> Maybe we can get some more user-friendly metadata e.g. `DisplayName`

Definitely!